### DISCUSSION TITLE
Pass variable to useAtom()

### DISCUSSION THREAD

**Discussion by mrks1337** [pmndrs/jotai#1927] [Q&A]
I have the following component where I get an `id` from `useParam` hook:

    import React from 'react';
    import { BaseProps } from 'core/model';
    import { ProgressOverlayFallback } from 'common/components/fallback/progress-overlay-fallback.component';
    import { useAtom } from 'jotai';
    import { integratorDetailAtom } from './integrator-detail.state';
    import { useParams } from 'react-router-dom';
    
    export type IntegratorDetailProps = BaseProps;
    
    export type UrlParamsIntegratorEdgeId = {
      integratorId: string;
    };
    
    export const IntegratorDetail: React.FC<IntegratorDetailProps> = () => {
      const { integratorId } = useParams<UrlParamsIntegratorEdgeId>(); // pass `integratorId` to useAtom
      const [{ isLoading }] = useAtom(integratorDetailAtom);
      return isLoading ? <ProgressOverlayFallback /> : <>Integrator Detail Pod</>;
    };

I want to use `integratorId` now in my `integratorDetailAtom` so I can include it in my `queryFn`:

    import { atom } from 'jotai';
    import { atomsWithQuery } from 'jotai-tanstack-query';
    import { apiKeys } from 'core/api';
    import { getIntegratorDetail } from './api';
    
    // TODO: replace `integratorEdgeIdAtom` with useState from parent tp pass to queryFn()
    const integratorEdgeIdAtom = atom(
      location.pathname.substring(location.pathname.lastIndexOf('/') + 1)
    );

    export const [, integratorDetailAtom] = atomsWithQuery(get => ({
      queryKey: [apiKeys.integrator.getDetail, get(integratorEdgeIdAtom)],
      queryFn: async ({ queryKey: [, id] }) => getIntegratorDetail(id), // here
    }));

I tried it with `useAtom(integratorDetailAtom(integratorId))` but this does not work and I can now wrap my head how to do this.

**Comment by dai-shi** [comment]
Atoms defined globally can't take params in general.

Here are some ideas:
- Make your source data as atoms. This isn't possible if the data is from other libs out of control.
- Create a new atom in render. We have to be careful to follow React coding rules. It's also tricky to pass the atom down the tree (context or atoms-in-atom).
- Use `atomFamily`. Be careful for memory leaks.
- Use `jotai-molecules`. The API has learning curve.
